### PR TITLE
fix(shim): Fix PS1 shim error when in different drive in PS7

### PR DESCRIPTION
**Pull request by niheaven**
#### Description

Fix PS1 shim's content switch, change from `^(.\\[\w]:).*$` to `^(\.\\)?\w:.*$`. Also enclose shim content with @() for a clear look.

#### Motivation and Context

PS1 shim went wrong in PowerShell 7 if target is in a different drive other than shim dir. This is introduced by `Resolve-Path -Relative` cmdlet's output.

In PS5:
image

In PS7:
image

Other changes are pure refactoring (enclosed raw code in `@()`)

#### How Has This Been Tested?

Run `Invoke-Pester ".\test\Scoop-Core.Tests.ps1"`

Before:
image

After:
image

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

**Comment by niheaven** [comment]
@rashil2000 Could you please check enclosed code? I do this carefully and think they haven't changed 😄

File: lib/core.ps1 [hunk 1]
```diff
@@ -577,90 +577,114 @@ function shim($path, $global, $name, $arg) {
...
-if command -v pwsh.exe &> /dev/null; then
-    pwsh.exe -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"
-else
-    powershell.exe -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"
-fi" | Out-File $shim -Encoding ASCII
+        @(
+            '#!/bin/sh',
+            "# $resolved_path",
+            'if command -v pwsh.exe &> /dev/null; then',
+            "    pwsh.exe -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"",
```
Comment (rashil2000) [inline lib/core.ps1:639]:
```suggestion
            "    pwsh -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"",
```

File: lib/core.ps1 [hunk 2]
```diff
@@ -577,90 +577,114 @@ function shim($path, $global, $name, $arg) {
...
+            'else',
+            "    powershell.exe -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"",
```
Comment (rashil2000) [inline lib/core.ps1:641]:
```suggestion
            "    powershell -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"",
```

**Review by rashil2000 — CHANGES_REQUESTED** [review]

**Comment by rashil2000** [comment]
Only one minor thing: Some places use `''` to enclose strings, while some use `""`. I know `""` is used only in places that have string literals, but I think we should try to be consistent instead. Mixing them up reduces readability.

File: lib/core.ps1 [hunk 3]
```diff
@@ -577,90 +577,114 @@ function shim($path, $global, $name, $arg) {
...
 
         warn_on_overwrite $shim $path
-        "#!/bin/sh
-# $resolved_path
-MSYS2_ARG_CONV_EXCL=/C cmd.exe /C `"$resolved_path`" $arg `"$@`"" | Out-File $shim -Encoding ASCII
+        @(
+            '#!/bin/sh',
+            "# $resolved_path",
+            "MSYS2_ARG_CONV_EXCL=/C cmd.exe /C `"$resolved_path`" $arg `"$@`""
+        ) -join "`r`n" | Out-File $shim -Encoding ASCII
```
Comment (rashil2000) [inline lib/core.ps1:590]:
```suggestion
        ) -join "`n" | Out-File $shim -Encoding ASCII
```

File: lib/core.ps1 [hunk 4]
```diff
@@ -577,90 +577,114 @@ function shim($path, $global, $name, $arg) {
...
-fi" | Out-File $shim -Encoding ASCII
+        @(
+            '#!/bin/sh',
+            "# $resolved_path",
+            'if command -v pwsh.exe &> /dev/null; then',
+            "    pwsh.exe -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"",
+            'else',
+            "    powershell.exe -noprofile -ex unrestricted -command `"& '$resolved_path' $arg $@;exit \`$lastexitcode`"",
+            'fi'
+        ) -join "`r`n" | Out-File $shim -Encoding ASCII
```
Comment (rashil2000) [inline lib/core.ps1:643]:
```suggestion
        ) -join "`n" | Out-File $shim -Encoding ASCII
```

File: lib/core.ps1 [hunk 5]
```diff
@@ -577,90 +577,114 @@ function shim($path, $global, $name, $arg) {
...
 
         warn_on_overwrite $shim $path
-        "#!/bin/sh
-# $resolved_path
-java -jar `"$resolved_path`" $arg `"$@`"" | Out-File $shim -Encoding ASCII
+        @(
+            '#!/bin/sh',
+            "# $resolved_path",
+            "java -jar `"$resolved_path`" $arg `"$@`""
+        ) -join "`r`n" | Out-File $shim -Encoding ASCII
```
Comment (rashil2000) [inline lib/core.ps1:656]:
```suggestion
        ) -join "`n" | Out-File $shim -Encoding ASCII
```

File: lib/core.ps1 [hunk 6]
```diff
@@ -577,90 +577,114 @@ function shim($path, $global, $name, $arg) {
...
 
         warn_on_overwrite $shim $path
-        "#!/bin/sh
-# $resolved_path
-python `"$resolved_path`" $arg `"$@`"" | Out-File $shim -Encoding ASCII
+        @(
+            '#!/bin/sh',
+            "# $resolved_path",
+            "python `"$resolved_path`" $arg `"$@`""
+        ) -join "`r`n" | Out-File $shim -Encoding ASCII
```
Comment (rashil2000) [inline lib/core.ps1:669]:
```suggestion
        ) -join "`n" | Out-File $shim -Encoding ASCII
```

File: lib/core.ps1 [hunk 7]
```diff
@@ -577,90 +577,114 @@ function shim($path, $global, $name, $arg) {
...
 
         warn_on_overwrite $shim $path
-        "#!/bin/sh
-# $resolved_path
-`"$resolved_path`" $arg `"$@`"" | Out-File $shim -Encoding ASCII
+        @(
+            '#!/bin/sh',
+            "# $resolved_path",
+            "`"$resolved_path`" $arg `"$@`""
+        ) -join "`r`n" | Out-File $shim -Encoding ASCII
```
Comment (rashil2000) [inline lib/core.ps1:687]:
```suggestion
        ) -join "`n" | Out-File $shim -Encoding ASCII
```

**Review by rashil2000 — CHANGES_REQUESTED** [review]
Changes to bash shim due to #4616

**Comment by niheaven** [comment]
> Only one minor thing: Some places use `''` to enclose strings, while some use `""`. I know `""` is used only in places that have string literals, but I think we should try to be consistent instead. Mixing them up reduces readability.

Well, the two different quotes are introduced by PowerShell formatter, I'll change them all to `"` for better readability.

File: lib/core.ps1 (hunk 1 above)
Comment (niheaven) [reply lib/core.ps1:639]: Are there some benefits to trim `.exe`? I'm not sure the difference of using or not the extension of exe file.

**Comment by niheaven** [comment]
It should be done.

**Review by rashil2000 — APPROVED** [review]

**Comment by niheaven** [comment]
@rashil2000 Could you please add some tests for newly added shims? e.g., `jar`, `python` or other types.

**Comment by rashil2000** [comment]
Sure!

But I have never written tests for PowerShell before. Could you point at some examples in the codebase I can look at?

**Pull request merged**